src/dsl.jl

```diff
@@ -298,149 +266,114 @@ end
-        ivexpr.args[1] = Symbol("@", "parameters")
-    else
-        ivs = (DEFAULT_IV_SYM,)
-        ivexpr = :($(DEFAULT_IV_SYM) = default_t())
-    end
-    tiv = ivs[1]
-    sivs = (length(ivs) > 1) ? Expr(:vect, ivs[2:end]...) : nothing
-    all_ivs = (isnothing(sivs) ? [tiv] : [tiv; sivs.args])
-
-    if haskey(options, :combinatoric_ratelaws)
-        combinatoric_ratelaws = options[:combinatoric_ratelaws].args[end]
-    else
-        combinatoric_ratelaws = true
-    end
-
-    # Collect species and parameters, including ones inferred from the reactions.
-    declared_syms = Set(Iterators.flatten((parameters_declared, species_declared,
-        variables_declared)))
-    species_extracted, parameters_extracted = extract_species_and_parameters!(
-        reactions, declared_syms; requiredec)
-
-    # Reads equations (and infers potential variables).
-    # Excludes any parameters already extracted (if they also was a variable).
-    declared_syms = union(declared_syms, species_extracted)
-    vars_extracted, add_default_diff, equations = read_equations_options(
-        options, declared_syms, parameters_extracted; requiredec)
-    variables = vcat(variables_declared, vars_extracted)
-    parameters_extracted = setdiff(parameters_extracted, vars_extracted)
-
-    # Creates the finalised parameter and species lists.
-    species = vcat(species_declared, species_extracted)
-    parameters = vcat(parameters_declared, parameters_extracted)
-
-    # Create differential expression.
-    diffexpr = create_differential_expr(
-        options, add_default_diff, [species; parameters; variables], tiv)
-
-    # Reads observables.
-    observed_vars, observed_eqs, obs_syms = read_observed_options(
-        options, [species_declared; variables], all_ivs; requiredec)
-
-    # Checks for input errors.
-    (sum(length.([reaction_lines, option_lines])) != length(ex.args)) &&
-        error("@reaction_network input contain $(length(ex.args) - sum(length.([reaction_lines,option_lines]))) malformed lines.")
-    any(!in(opt_in, option_keys) for opt_in in keys(options)) &&
-        error("The following unsupported options were used: $(filter(opt_in->!in(opt_in,option_keys), keys(options)))")
-    forbidden_symbol_check(union(species, parameters))
-    forbidden_variable_check(variables)
-    unique_symbol_check(union(species, parameters, variables, ivs))
+    sps_inferred, ps_pre_inferred = extract_sps_and_ps(reactions, syms_declared; requiredec)
+    vs_inferred, diffs_inferred, equations = read_equations_option!(diffsexpr, options,
+        union(syms_declared, sps_inferred), tiv; requiredec)
+    ps_inferred = setdiff(ps_pre_inferred, vs_inferred, diffs_inferred)
+    syms_inferred = union(sps_inferred, ps_inferred, vs_inferred, diffs_inferred)
+    all_syms = union(syms_declared, syms_inferred)
+    obsexpr, obs_eqs, obs_syms = read_observed_option(options, ivs,
+        union(sps_declared, vs_declared), all_syms; requiredec)
```

Comment:
Step 3: Read all reactions and equations. Finds all symbols that are inferred as something.